Issue774 check fields

einola opened the pull request:
# Checks whether model fields have physically reasonable values
Fixes parts of #774

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR adds a checkFields function to the ModelComponent class. This means that all derived classes, including all modules, can use it to check the validity of the model fields. The function checks if a given field is in a given range, i.e. min < field[i] < max for all ocean points.

This PR includes two example implementations for the ERA5Atmosphere and TOPAZOcean modules, as well as for PrognosticData. It also includes some testing of the new functionality.

When field values outside the prescribed bounds are found, the function throws a runtime_error, which is caught by the Iterator class. When the exception is caught, the iteration is stopped and the iterator returns a non-zero exit code to the calling function. This means all netCDF files should be closed properly and a final restart file should be written out. 

I also used boost/stacktrace functionality to print out the stack trace when the exception is caught. This does not work on all platforms, but the code should detect functioning and non-functioning platforms and compile accordingly. A more portable approach would be preferrable, but I didn't find one.

At the moment, the stack trace doesn't print file name and line number of the offending call, only the function name. This can by fixed by not linking to nextsimdglib, but directly to the object files themselves.

This PR addresses points 2 and 3 of issue #774 

---
# Test Description

The additional code is tested in two ways, through the realistic integration test.
 1. The integration test now includes test_fields = true for the ERA5Atmosphere and TOPAZOcean modules
 2. I added a second integration test run with erroneous input data and a test to check that the model returns a non-zero exit code.

---
# Documentation Impact

Addresses points 2 and 3 of issue #774.

We should have a developer's guide to writing modules and include a description of how to check module fields. But this needs to be in a separate issue.

---
# Other Details

N/A

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

einola committed:
Change IIceOceanHeatFlux and IFreezingPoint references to pointers

I don't understand why, but using Module::getImplementation<Interface>
directly reset nOcean, oceanIndex, and p_oceanMaskH. But if I first get
a pointer to the implementation everything works. A bonus is that we can
call tryConfigure for these modules in a transparent way.

einola committed:
Add a function to check fields to ModelComponent

The function checkFields will check all fields listed in fieldsToCheck,
using overElements and checkFieldsElement (which does all the work).
checkFieldsElement checks if the given fields are within the bounds
listed in fieldsToCheck and throws a runtime_error if the check fails.
Each model component can then populate it's own fieldsToCheck vector and
call checkFields to check it's internal logic.

einola committed:
Minor cleaning of ModelComponent

nOcean should be set to zero before entering the loop with ++nOcean. I
also think the function oceanMask() should be called getOceanMask(),
since we already have setOceanMask().

einola committed:
Implement checks for TOPAZOcean

Use the checkFields functionality of ModelComponent to check the input
fields read in from the TOPAZ files.

einola committed:
Implement checks for ERA5Atmosphere

Use the checkFields functionality of ModelComponent to check the input
fields read in from the ERA5 files.

einola committed:
Minor cleaning of checks for TOPAZOcean

Change the test to be < and > and not <= and >=. This works better for
everything except mld, where we have to make sure mld > 0 by saying mld
>= 1e-4. Fix a bug to ensure SSS >= 0. Some clang_format changes.

einola committed:
Unconditinonally check basic prognostics in PrognosticData

We may want to make these checks conditional later, but I don't think so
. It's too bad we can't check velocity here, but that'll have to be done
 in the dynamics modules.

einola committed:
Add checkFields to ERA5Atmosphere.cpp

I for got to do that before.

einola committed:
Handle exception thrown by checkFieldsElement

This adds a try-catch clause in iterator, which then exits the main
model loop with a non-zero return code, printing the boost-generated
stack trace to the command line. This way, the model exits gracefully,
closing all files and creating a restart file, but returns a non-zero
return code.

The only problem here is that boost stacktrace may not always be
implemented, so the code may not compile on all machines. We need to
discuss how to address this, because the stack trace is a very usefull
thing to have.

It would also be nice to have the i and j index and not only the linear
index. And it would be nice to have the file name and line number, not
just "in [...] libnextsimlib.so".

einola committed:
Skip the backtrace code when the functionality is not available

Boost has a set of CPP keys which together tell you if the backtrace
code is there. I copied this into stackTraceException.hpp, so now the
code compiles without backt trace functionality if this is not present
in boost.

einola committed:
Add tests

Use the realistic integration test to test error handling in two ways.
Add "check_errors" to integration_test.cfg, and rerun the integration
test with erroneous input data. This second run should fail with a
non-zero exit code.

einola committed:
Also run error handling test on the macos image

I forgot to add that on my last commit.

einola committed:
Echo a message that the error handling test was successful

It's a nice thing to do!

timspainNERSC commented on core/src/PrognosticData.cpp:
```diff
@@ -51,6 +51,13 @@ void PrognosticData::configure()
     tryConfigure(iceGrowth);
+
+    // clang-format off
+    fieldsToCheck.push_back({"m_conc",   &m_conc,   std::make_pair(0, 1+1e4)});
+    fieldsToCheck.push_back({"m_damage", &m_damage, std::make_pair(0, 1+1e-4)});
+    fieldsToCheck.push_back({"m_snow",   &m_snow,   std::make_pair(0, 10)});
+    fieldsToCheck.push_back({"m_thick",  &m_thick,  std::make_pair(0, 50)});
```
Use the names names in `gridNames.hpp`.
Use the `ModelArrayRef` system, since the model has infrastructure for mapping names to `ModelArray` fields.
Use an initializer list, rather than `std::make_pair`.

timspainNERSC reviewed (changes requested):
Rather than turning on checks per-module, I think it would be better to take a list of labels for `ModelArrayStore` (that is the same ones used in `ConfigOutput.field_names`) to be checked. This would allow the fields that are checked in any particular run to be set at run time, without hard coding the set of fields in each `ModelComponent`.

einola committed:
Minor cleaning suggested by Tim on the PR

Use the names names in gridNames.hpp.
Use an initializer list, rather than std::make_pair.

einola committed:
Also check m_tice

This required some extra logic to go into the 3rd dimension.

einola committed:
Limit damage after dynamics

Otherwise, we get values like 1+1e-9 > 1, which is no good. It seems to
come from the projection from Gauss points back to DG space. Ideally,
we'd like to limit only once, but I'm not sure if that's possible.

I also changed avgU.zero() to avgU.setZero(), although I'm not sure what
the practical difference is.

einola committed:
Test all registered arrays in PrognosticData

Following the suggestion from Tim, I implemented functionality to test
all registered ModelArrays in PrognosticData. I'll then make this
testing optional, or at least make it optional to test some arrays.

einola committed:
Change include statements for StackTraceException.hpp

The refactoring apparently didn't work. It compiles on my machine, but
not in the CI.

einola committed:
Checks and modifications so that the column test works

We shouldn't check ModelArrays that aren't initialised (like
IO_STRESS_X), so I added a bit hackish check for this. Turns also out
that Q_PEN_SW can be quite high.

einola committed:
Don't put unused arrays into fieldsToCheck

To do this, I moved the loop over getStore().getAllData() return to the
end of setData. At this point we've called setData on all model
components and modules, so a data().rows() call returns omething useful.
Removed this check from checkFields and did some minor cleaning there.

einola committed:
User defined checks

Allows the user to decide if PrognosticData checks or not and allows her
to supply a list of variables to check. The list is still the TextTag
names, though, not the output names listed in ProtectedArrayNames.ipp
and SharedArrayNames.ipp

einola committed:
Help prints "true" and "false" instead of 1 and 0 for bools

Because it's prettier.

einola committed:
Clang-format fix

Just formatting

einola committed:
Chang to using output names for checking fields

The internal logic still uses the TextTag names, but I use externalNames
and a reverse-lookup map of that to map between those and the output
names. I also added debug.fields = ALL to integration_test.cfg.

einola commented:
Very good suggestions and I think everything is in place now.

I'm keeping the ability to do per-module error checking. I think this can be very useful for debugging a model that is too big or the run is too long to put through a debugger.

einola committed:
Split the setting of fieldsToCheck into a separate function

It's just a bit cleaner.

einola committed:
Minor formatting

No functionality changes

einola committed:
Fix clang-format problems

They don't show up on my machine, but the CI isn't happy.

einola committed:
Remove boost stacktrace exception handling

It doesn't work very well, and since PrognosticData does most of the
error handling a simple pfx +  to the error message is enough.

TomMelt commented on test/run_test_error_handling_test.sh:
```diff
@@ -0,0 +1,33 @@
+restart_file=init_25km_NH.nc
+era5_file=25km_NH.ERA5_2010-01-01T000000_test_data.nc
+topaz_file=25km_NH.TOPAZ4_2010-01-01T000000_test_data.nc
+out_file=out.integration_test.nc
```
do these do anything? or are they more comments?

TomMelt commented on core/src/ModelComponent.cpp:
```diff
@@ -122,6 +123,42 @@ ModelArray ModelComponent::mask(const ModelArray& data)
+        for (layerToCheck = 0; layerToCheck < nLayers; ++layerToCheck)
+            overElements(std::bind(&ModelComponent::checkFieldsElement, this, std::placeholders::_1,
+                             std::placeholders::_2),
+                tst);
```
Not a problem, but just curious. Is there a reason to use `bind` over lambdas?

TomMelt commented on core/src/include/PhysicalBounds.ipp:
```diff
@@ -0,0 +1,67 @@
+{ "H_ICE_cell", {0, 50} }, // Ice thickness, cell average, m
+{ "C_ICE0", {0, 1} }, // Ice concentration
+{ "H_SNOW_cell", {0, 10} }, // Snow depth, cell average, m
+{ "T_ICE0", {-100, 0} }, // Ice temperature, ˚C
```
Also not a big deal, but just raising unicode chars could cause issues for some tools (even in comments). No need to change just raising awareness

TomMelt reviewed (approved):
Thanks @einola this will be really helpful to reduce user error.

TomMelt commented:
@einola , not exactly what you asked for i.e., stack trace, but it is possible to get the filename and line number via macros which should be a truly cross-platform solution e.g.,:

```cpp
std::cout << __FILE__ << " " << __LINE__ << "\n";
```

einola committed:
Remove unused variables from test script

Just to make the script cleaner.

einola committed:
Two minor bug fixes to the physics

These became apparent when pushing the model using the field-checking
implemented in this branch, so I'm including them here. One sets qio to
zero if there's no ice and the other sets the tice to the freezing
point of ice when there's no ice in ThermoIce0. In the Winton model,
tice was already correctly set.

timspainNERSC commented on core/src/ModelComponent.cpp:
```diff
@@ -1,13 +1,13 @@
 #include "include/ModelComponent.hpp"
-
 #include "include/MissingData.hpp"
+#include <cmath>
```
I've been laying out the headers in source files as:
"HeaderForThisClass.hpp"
[blank line]
"OtherNextSIMHeader.hpp"
…
[blank line]
<system_headers>
…

This system groups the headers together and avoids clang-format from complaining about alphabetization.
Could you revert to that layout here? (Add a blank line between each of the headers)

timspainNERSC commented on core/src/PrognosticData.cpp:
```diff
@@ -88,6 +110,51 @@ void PrognosticData::setData(const ModelState::DataMap& ms)
+            fieldsToCheck.push_back(
+                { pfx + ": " + reverseNames.at(x.first), x.second, bounds.at(x.first) });
+        }
+    }
```
This should not be in `PrognosticData`. There should probably should be an independent error checking class.

einola replied:
I guess the point is that this is not a part of the core functionality of ``PrognosticData``, which I agree with. But this check should only be done once and only in ``PrognosticData``, so shouldn't it then be confined to ``PrognosticData`` itself?

timspainNERSC commented on core/src/include/ConfigurationHelp.hpp:
```diff
@@ -38,6 +38,14 @@ class ConfigurationHelp {
+            return { "true" };
+        else
+            return { "false" };
+    }
```
This is already a feature of the standard library, see [Converting bool to text in C++](https://stackoverflow.com/questions/29383/converting-bool-to-text-in-c) on StackOverflow.

timspainNERSC commented on core/src/include/ModelArrayReferenceStore.hpp:
```diff
@@ -72,6 +72,18 @@ class ModelArrayReferenceStore {
         return dataMap;
     }
 
+    // For a given field name return the ModelArray pointer pointing to the data
+    const ModelArray* getArrayRef(const std::string& field)
+    {
+        if (const auto returnValue = storeRW.find(field); returnValue != storeRW.end())
+            return returnValue->second;
+
+        if (const auto returnValue = storeRO.find(field); returnValue != storeRO.end())
+            return returnValue->second;
+
+        return nullptr;
+    }
+
```
`storeRW` and `storeRO` are already `std::map`s from the `std::string name`. You don't need to use `find()`.

einola replied:
Yes, but ``storeRW``and ``storeRO`` may not contain field. I now use nested ``try`` instead and also try and catch in the calling function. I'm not sure which is better.

timspainNERSC commented on core/src/include/PhysicalBounds.ipp:
```diff
@@ -0,0 +1,67 @@
+/*!
+ * @file PrognosticData.cpp
```
Correct the name of the file.

timspainNERSC commented on core/src/include/PhysicalBounds.ipp:
This could be a class (probably derived from some `std::map`) that returns the limits `std::pair` give a name (or preferably `TextTag`).

timspainNERSC commented on core/src/include/PrognosticData.hpp:
Move these additions to a new error checking class.

timspainNERSC commented on dynamics/src/include/BrittleCGDynamicsKernel.hpp:
```diff
@@ -114,8 +114,8 @@ template <int DGadvection> class BrittleCGDynamicsKernel : public CGDynamicsKern
 
-        avgU.zero();
-        avgV.zero();
+        avgU.setZero();
+        avgV.setZero();
```
Why change this? `zero()` works perfectly well, and is part of the `cgVector` interface, as opposed to the parent `Eigen::Matrix` interface.

einola replied:
Why indeed? I was reading the Eigen::Matrix interface for something else and got confused. Reverted.

timspainNERSC commented on physics/src/modules/AtmosphereBoundaryModule/ERA5Atmosphere.cpp:
```diff
@@ -49,6 +54,20 @@ void ERA5Atmosphere::configure()
+
+    if (doChecks) {
+        // clang-format off
+        fieldsToCheck.push_back({pfx+": tair",  &tair,  {  -100, 80}});
+        fieldsToCheck.push_back({pfx+": tdew",  &tdew,  {  -100, 80}});
+        fieldsToCheck.push_back({pfx+": pair",  &pair,  { 700e2, 1100e2}});
+        fieldsToCheck.push_back({pfx+": sw_in", &sw_in, {    -1, 2e3}});
+        fieldsToCheck.push_back({pfx+": lw_in", &lw_in, {     0, 1e3}});
+        fieldsToCheck.push_back({pfx+": wind",  &wind,  {  -0.1, 150}});
+        fieldsToCheck.push_back({pfx+": uwind", &uwind, {  -150, 150}});
+        fieldsToCheck.push_back({pfx+": vwind", &vwind, {  -150, 150}});
```
I think this would look neater if you merged in a initializer list literal. There would at least be only one function call. But this one is a matter of taste.

timspainNERSC commented on physics/src/modules/IceOceanHeatFluxModule/BasicIceOceanHeatFlux.cpp:
```diff
@@ -27,6 +27,8 @@ void BasicIceOceanHeatFlux::updateElement(size_t i, const TimestepTime& tst)
     // Use the timestep length as the relaxation time scale
     if (cice[i] > 0.) {
         qio[i] = doOne(tf[i], sst[i], mlBulkCp[i], tst.step.seconds());
+    } else {
+        qio[i] = 0.;
```
This feels like it should be in its own bug fix PR.

einola replied:
Now in #789

timspainNERSC commented on physics/src/modules/IceThermodynamicsModule/ThermoIce0.cpp:
```diff
@@ -87,6 +87,7 @@ void ThermoIce0::calculateElement(size_t i, const TimestepTime& tst)
     if (hice[i] == 0. || cice[i] == 0.) {
         deltaHi[i] = 0.;
         snowToIce[i] = 0.;
+        tice.zIndexAndLayer(i, 0) = freezingPointIce;
```
This should be in a bug fix PR, probably along with `BasicIceOceanHeatFlux:30`.

einola replied:
Now in #789

timspainNERSC commented on physics/src/modules/OceanBoundaryModule/TOPAZOcean.cpp:
```diff
@@ -47,7 +50,25 @@ void TOPAZOcean::configure()
+
+    if (doChecks) {
+        // clang-format off
+        fieldsToCheck.push_back({pfx+": sstExt", &sstExt, {  -5, 50}});
+        fieldsToCheck.push_back({pfx+": sssExt", &sssExt, {   0, 50}});
+        fieldsToCheck.push_back({pfx+": mld",    &mld,    {1e-3, 11e3}});
+        fieldsToCheck.push_back({pfx+": u",      &u,      {  -5, 5}});
+        fieldsToCheck.push_back({pfx+": v",      &v,      {  -5, 5}});
+        fieldsToCheck.push_back({pfx+": ssh",    &ssh,    {  -5, 5}});
```
See the comment on `ERA5Atmosphere.cpp`.

timspainNERSC commented on physics/src/modules/OceanBoundaryModule/TOPAZOcean.cpp:
```diff
@@ -97,9 +120,6 @@ void TOPAZOcean::setData(const ModelState::DataMap& ms)
-{
-    tf[i] = Module::getImplementation<IFreezingPoint>()(sss[i]);
-}
+void TOPAZOcean::updateTf(size_t i, const TimestepTime& tst) { tf[i] = (*pFreezingPoint)(sss[i]); }
```
I'd rather this continued to use the `Module` infrastructure.

einola replied:
Ah, yes. I should have asked you about this one. See commit message for bb47c1e1

> I don't understand why, but using Module::getImplementation<Interface>
> directly reset nOcean, oceanIndex, and p_oceanMaskH. But if I first get
> a pointer to the implementation everything works. A bonus is that we can
> call tryConfigure for these modules in a transparent way.

timspainNERSC commented on physics/src/modules/OceanBoundaryModule/include/TOPAZOcean.hpp:
```diff
@@ -43,14 +45,19 @@ class TOPAZOcean : public IOceanBoundary, public Configured<TOPAZOcean> {
     HField sssExt;
 
     SlabOcean slabOcean;
+
+    IIceOceanHeatFlux* pIOHeatFlux;
+    IFreezingPoint* pFreezingPoint;
```
These shouldn't be raw pointers. In fact, I don't think they should be here at all. Follow the method in the other ocean modules, and use the `Module` infrastructure.

einola replied:
I copied what you did in ERA5Atmosphere with the IFluxCalculation reference #cargocultcoding I tried a bit with smart pointers, but I wasn't smart enough to make it work.

timspainNERSC reviewed (changes requested):
Several changes to make (and possibly discuss).

einola committed:
Use std::boolalpha to format true and false on output

I was using an if-clause

einola committed:
Fix header file listing

Headerfiles should be grouped by empty spaces into "header for this
class", "other nextsim headers", and "system headers".

einola committed:
Rewrite getArrayRef to use try

It now throws std::out_of_range if the field is in neither storeRW nor
storeRO. I then catch this in PrognosticData.cpp.

einola committed:
Correct the name in PhysicalBounds.ipp header

No code changes.

einola committed:
Revert useless change to BrittleCGDynamicsKernel.hpp

It was a commit of mine after exploring, and not understanding the
cgVector interface.

einola committed:
Turn PhysicalBounds.ipp into a class

It wraps an std::map<TextTag, std::pair<double,double> of all the
physical bounds. Use The class in all fieldsToCheck.push_back's.

einola committed:
Move addBounds into constructor

The CI didn't like that I hadn't defined addBounds before using it. So I
moved it into the class constructor.

einola commented:
I think this should be it except
 - The use of raw pointers - which I don't know how to fix
 - Should setFieldsToCheck be in a separate class? I'm not sure

Maybe a discussion in meat space would help.

einola committed:
Use unique_ptr instead of raw pointers to modules

Reading the code, I realised that IceGrowth.cpp uses unique_ptr to
reference Module instances - so I copied that approach. It only requires
minor changes.

einola committed:
Split checking into a separate class

This puts all the checking into the CheckingModelComponent class.
PrognosticData, IAtmosphereBoundary, and IOceanBoundary now inherit from
this class, instead of ModelComponent, to activate the checking.

einola committed:
Use setFieldsToCheck in ERA5 and TOPAZ boundary classes

Instead of a manual push_back of field names and bounds, use the
functionality in setFieldsToCheck. This is possible, since all the
fields these routines should check are registered in the
ModelArrayReference store.

To get this to work, I had to move the check
for empty fields from setFieldsToCheck to checkFields, since it's not
guarantied anymore that all fields have been set/resized when
setFieldsToCheck is called. I also moved several static constants in
ERA5Atmosphere.cpp and TOPAZOcean.cpp into the respective name spaces to
avoid conflict with constants named in PrognosticData.cpp. Finally, I
moved the call to setFieldsToCheck into the configure() function of the
respective classes. Logically, it belongs there, and it can now be
called from there as well, because it doesn't check if the fields have
been set/resized.

einola commented:
Using std::unique_ptr and a separate class for consistency checking. Can you have another look @timspainNERSC

timspainNERSC commented on core/src/CheckingModelComponent.cpp:
```diff
@@ -0,0 +1,100 @@
+/*!
+ * @file ModelComponent.cpp
```
Change the file string to match the file name

timspainNERSC commented on core/src/CheckingModelComponent.cpp:
```diff
@@ -0,0 +1,100 @@
+    }
+}
+
+} /* namespace Nextsim */
```
Github marks the lack of a final newline with a big red symbol, so presumably its important. Add this.

timspainNERSC commented on core/src/include/CheckingModelComponent.hpp:
```diff
@@ -0,0 +1,49 @@
+/*!
+ * @file ModelComponent.cpp
```
Change the file string to match the file name

timspainNERSC commented on core/src/CheckingModelComponent.cpp:
```diff
@@ -0,0 +1,100 @@
+            throw std::logic_error(
+                "ModelComponent::checkFields expected a field with 2 or 3 dimensions.\n");
+
+        // layerToCheck must be in scope for checkFieldElement as well
```
The checking be better performed using `Eigen` library functions, such as `isnan()`, `max()`, `maxCoeff(posn)`, rather than trying to hack around `overElements()`.

timspainNERSC commented on core/src/PrognosticData.cpp:
```diff
@@ -51,6 +62,10 @@ void PrognosticData::configure()
     tryConfigure(pDynamics);
 
     tryConfigure(iceGrowth);
+
+    if (getConfiguration(keyMap.at(CHECKFIELDS_KEY), checkFieldsDefault)) {
+        setFieldsToCheck(getConfiguration(keyMap.at(FIELDNAMES_KEY), fieldNamesDefault), pfx);
+    }
```
Put the common code and configuration keys into the `CheckingModelComponent` class.

einola replied:
I don't know how to do that.

``checkFieldsDefault`` and ``fieldNamesDefault`` are unique to the child classes ``PrognosticData``, ``ERA5Atmosphere``, or ``TOPAZOcean``, or other unimplemented uses of ``CheckingModelComponent``. The default field list should depend on the child class, and whether to check or not should be true for ``PrognosticData`` and false for everyone else.

``keyMap`` is also unique to the child class because the string can have different prefixes, e.g. ``ERA5Atmosphere.check_fields`` vs. ``TOPAZOcean.check_fields``.

So, I agree that there's too much repeated code, but I don't see how to reduce it meaningfully.

timspainNERSC commented on core/src/PrognosticData.cpp:
```diff
@@ -177,13 +194,29 @@ ModelState PrognosticData::getStateRecursive(const OutputSpec& os) const
 }
 
-PrognosticData::HelpMap& PrognosticData::getHelpText(HelpMap& map, bool getAll) { return map; }
+PrognosticData::HelpMap& PrognosticData::getHelpText(HelpMap& map, bool getAll)
+{
+
+    map["Debug"] = {
+        { fieldNamesKey, ConfigType::STRING, {}, fieldNamesDefault, "",
+            "Comma separated, space free list of fields to be checked if check_fields is true. "
+            "The special value \""
+                + all + "\" will output all available fields." },
+        { checkFieldsKey, ConfigType::BOOLEAN, { "true", "false" },
+            ConfigurationHelp::toString(checkFieldsDefault), "",
+            "Switch to check if the fields listed in field_names are within a reasonable "
+            "physical range and not NaN." },
+    };
+    return map;
+}
+
```
Put the common help test pertaining to checking into the `CheckingModelComponent` class.

timspainNERSC commented on core/src/Iterator.cpp:
```diff
@@ -37,16 +37,26 @@ TimePoint Iterator::parseAndSet(const std::string& startTimeStr, const std::stri
     for (auto t = startTime; t < stopTime; t += timestep) {
         TimestepTime tsTime = { t, timestep };
-        iterant.iterate(tsTime);
+        try {
```
Why catch the exception at this level, just to pass an integer return value up to the next layer?

einola replied:
I did that so that the output buffers would be flushed, the file would be closed, and the model would write an end-of-run restart file, which may contain more interesting information. Not flushing the buffers before crashing is a significant problem in the model's current state.

But this can maybe be done more elegantly, e.g.

```
        } catch (const std::exception& e) {
            interant.stop(tsTime);
            throw;
        }
```

And then catch the exception at a higher level.

timspainNERSC commented on core/src/main.cpp:
```diff
@@ -54,11 +56,11 @@ int main(int argc, char* argv[])
         model.configure();
         // Run the Model
-        model.run();
+        return_code = model.run();
```
Since the exceptions are not actually being handled at the lower levels, if returning a return code is necessary, rather than just letting the unhandled exception escape main, do it handle the exception here.

timspainNERSC commented on core/src/CheckingModelComponent.cpp:
```diff
@@ -0,0 +1,100 @@
+void CheckingModelComponent::checkFieldsElement(size_t i, const TimestepTime& tst) const
+{
+    const auto& x = fieldsToCheck[numToCheck];
+    const double& value = std::get<1>(x)->zIndexAndLayer(i, layerToCheck);
+    const auto& bounds = std::get<2>(x);
```
Make this tuple into an actual `class` (or perhaps a `struct`).

timspainNERSC reviewed (changes requested):
I think this needs a fair bit of restructuring.

einola committed:
Use try-catch throughout and not return codes

See Tim's comment on PR #784. Now Iterator::run() catches the exception,
calls iterant.stop(t) and rethrows the exception. This is then caught by
Model::run(), which catches the exception, calls writeRestartFile() and
Finalizer::finalize(), and rethrows the exception. This is caught by
main, which logs the error using Logged::error, sets a return code and
continues as before.

I also changed Logged output defaults, so that Logged::error generates
output for both the log file and stdout/stderr.

We could return different codes for different types of errors, but this
is not implemented, and maybe not needed either.

einola committed:
Change fieldsToCheck into a vector of classes

See Tim's comment to PR #784. I replaced the vector of tuples with a
vector of class objects, where the class contains the same objects as
the tuple. The benefit of using a class is a simpler syntax, while still
being able to use brace initialisation.

einola committed:
Rewrite checkFields to use Eigen methods instead of overElements

See Tim's comment to PR #784. The code is much shorter and easier to
read now. The only drawback is that we need to mask out land values,
which requires copying the whole field.

einola committed:
Minor formatting changes

Fixes header comments and "no new line at the end of file" issue.

einola committed:
Merge remote-tracking branch 'origin/develop' into issue774_check_fields

# Conflicts:
#	core/src/include/ModelComponent.hpp
#	physics/src/modules/OceanBoundaryModule/TOPAZOcean.cpp
#	physics/src/modules/include/IAtmosphereBoundary.hpp
#	physics/src/modules/include/IOceanBoundary.hpp

einola committed:
Add bounds etc for new fluxes

This adds physical bounds for the newly calculated open water stress and
short-wave fluxes in open water and ice covered waters (through the
ice).

einola committed:
New masking which also works for Z-fields

I had to rewrite ModelComponent::mask so that it would handle NaNs and
an arbitrary fill value. I also added outputting array index instead of
the raw index and reduced code duplication.

einola committed:
Minor cleaning and less code repetitions

einola committed:
Common help pertaining to checking now in CheckingModelComponent

See Tim's comment to PR #784

einola committed:
Minor fixes to help text

There was a period missing for TOPAZOcean.field_names and
ERA5Atmosphere.field_names and the text for ERA5Atmosphere.check_fields
was printed twice.

einola committed:
Rename getHelpList

getCheckingHelpList is a more descriptive name.

einola commented:
Some solutions and a DRY-fail. Can you have a look @timspainNERSC ?

einola commented:
Ping @timspainNERSC

einola committed:
Merge remote-tracking branch 'origin/develop' into issue774_check_fields

# Conflicts:
#	core/src/ModelComponent.cpp
#	core/src/PrognosticData.cpp
#	core/src/include/ModelComponent.hpp

einola committed:
Small changes to get things to run and compile after merge

In ModelComponent.cpp, oceanMasked was renamed to getOceanMask, and I
needed to add C_ICE_DG and H_ICE_DG to PhysicalBounds.hpp and SharedArrayNames.ipp

einola committed:
Merge branch 'develop' into issue774_check_fields

# Conflicts:
#	dynamics/src/include/BrittleCGDynamicsKernel.hpp

einola committed:
Fix clang-formatting

Removed an empty line.

einola committed:
Merge branch 'develop' into issue774_check_fields

einola committed:
Merge branch 'develop' into issue774_check_fields

# Conflicts:
#	core/src/ModelComponent.cpp

timspainNERSC commented:
I think that everything that is not directly to do with the error checking should be in a separate PR (or possibly several), including adding the return codes and the exception handling.

timspainNERSC commented:
The more I look at the design of the error checking part, the less I like it. The bounds are all defined in a single file, mixing together dynamics and column physics fields. I still regret putting all the ModelComponent field names in one header. Let's not compound that mistake further.

I have a suggestion for an alternative design that would allow the individual ModelComponents to define their own limits, meaning that new implementations could use the error checking without having to expose their implementation specific data to a single big file.

- Add an upper and lower limit to `ModelArray`. The limit values default to ±inf. Add a `setLimits(double lower, double upper)` function to `ModelArray`.
- Add a `check()` function to `ModelArray` and `ModelArrayRef`, which checks the array for NaN and that the values are within the defined limits. This should use the `Eigen` functions available on `ModelArray::DataType`.
- The new class `CheckingModelComponent` could provide a pure virtual `checkFields()` function that checks the fields relevant to its domain of the model.
- A common way of configuring which fields to check could also be provided here, at the `CheckingModelComponent` level.
- The fields that will be checked or are available to check should be listed in the help text for that component, along with the limits and if they are configurable (hice shouldn't be, but perhaps ice stress could be).

einola committed:
Merge remote-tracking branch 'origin/develop' into issue774_check_fields

einola committed:
Minor changes to get the model to compile after the merge

Make sure to use Configured::getConfiguration, not just
getConfiguration, and getOceanMask(), instead of oceanMask().

einola committed:
Add physical bounds for surface temperature

T_SURF defined both in Protected and Shared

einola committed:
Revert changes to ModelComponent::mask

The original version doesn't handle NaNs well, so I wrote a new one ...
and then overwrote this again in the last merge. I've now reinstated my
rewritten version.

einola committed:
Upp the bounds on delta_hice

They were a bit too strict.

einola committed:
New CheckingModelComponent design

Mostly following Tim's review suggestions. The check itself now takes
place in ModelArray. Construction of a model array may initialise the
physically reasonable limits, but these can also be set via a setLimits
function. CheckingModelComponent now only provides booleans and a list
of variables to check, as well as a minimal checking function for
modules to call. Implementation examples are in IOceanBoundary and
TOPAZOcean, as well as IAtmosphereBoundary and ERA5Atmosphere. I
implemented a "global" check for PrognosticData, where all the variables
in the ModelArrayDataStore are checked. The check is more useful if the
modules have set the physical limits of their variables (otherwise these
are set to +/- infinity). But we always check for NaNs. The model fails
these tests because there are some uninitialised ModelArrays hanging
around.

einola committed:
Punctuation and grammar in ModelArray::checkLimits

Just changes in output strings.

einola committed:
A better way to handle the limits and fill value

The fillValue is now set in setLimits, and this is called by the
constructor.

einola committed:
Spelling and grammar

Just output strings and comments.

einola committed:
Updated limits

Some of these are quite high, so I will have to check better what's
going on.

einola commented:
Working on this branch, I found some problems with the way we pass evaporation and precipitation around. I'm relegating this to a separate PR.

einola committed:
Prepend function information to the error message

The user needs to program this in, but it's very useful to have. Also
minor change to the error message to clarify it.

einola commented:
Depends on #853

einola committed:
Merge branch 'develop' into issue774_check_fields

# Conflicts:
#	core/src/include/ModelComponent.hpp
#	physics/src/modules/include/IAtmosphereBoundary.hpp

einola committed:
Merge branch 'develop' into issue774_check_fields

# Conflicts:
#	core/src/modules/include/ProtectedArrayNames.ipp

einola committed:
Add checks on fields that are read in for ERA5

Not much more to say about that.

einola committed:
Fix merge

There was a straggling emp after the merge.

einola committed:
Tweak the bounds for a few variables

A longer run showed that this is needed.

einola committed:
Merge branch 'develop' into issue774_check_fields

einola committed:
Move checks for booleans into checkFields

Since boolCheckFields and boolCheckAll() are class members, then we
might as well do the check inside the checkFields() function. I also
removed the TimestepTime parameter from that function, since it was not
used anymore (I should have done this in one of the previous commits).

einola committed:
Bugfix for checkLimits

We need to have the same dimensions for the mask as for the field to
check. But only want to check the mean DG value, so it's ok.

einola committed:
Add a check_fields_fast functionality to PrognosticData

By default, we always check if ice thickness, concentration, and the
velocities are making sense.

einola committed:
Rename getOceanMask to oceanMask

It's called oceanMask in develop, but getOceanMask here. I don't know
how this happened, but I'm renaming it to oceanMask to stay the same as
develop.

einola committed:
Remove unused includes from ModelArray.cpp

Model.hpp was mysteriously added, and I also removed some others that
weren't being used any more.

einola committed:
Set penSW in ThermoWintonTemperature_test.cpp

It used to be set to zero by default, but now it defaults to NaN. Also,
some clang-formatting.

einola committed:
Set penSW in IceGrowth_test.cpp

It used to be set to zero by default, but now it defaults to NaN.

einola committed:
DummyDynamics sets velocities to zero

So that we don't trip the check_field_fast check in PrognosticData.

einola committed:
Merge remote-tracking branch 'origin/develop' into issue774_check_fields

# Conflicts:
#	physics/src/modules/OceanBoundaryModule/TOPAZOcean.cpp

einola committed:
Merge remote-tracking branch 'origin/develop' into issue774_check_fields

einola committed:
Merge remote-tracking branch 'origin/develop' into issue774_check_fields

# Conflicts:
#	physics/src/modules/OceanBoundaryModule/TOPAZOcean.cpp
#	physics/src/modules/OceanBoundaryModule/include/TOPAZOcean.hpp
#	physics/src/modules/include/IOceanBoundary.hpp

einola committed:
Merge branch 'develop' into issue774_check_fields

timspainNERSC commented on core/src/include/ModelArray.hpp:
```diff
@@ -113,6 +114,18 @@ class ModelArray {
      * Construct an unnamed ModelArray of Type::H
      */
     ModelArray();
+    /*!
+     * @brief Construct a ModelArray of the given type and name
+     *
+     * @param t The ModelArray::Type for the new object.
+     * @param bounds The physical lower and upper bounds for the new object. See lowerPhysicalLimit
+     * and upperPhysicalLimit.
+     */
+    ModelArray(const Type t, const std::pair<double, double>& bounds)
+        : ModelArray(t)
+    {
+        setLimits(bounds.first, bounds.second);
+    }
```
Suggest keeping a constructor with parameters `ModelArray(const Type t)`, which delegates to this constructor by supplying default limit arguments of `{-std::numeric_limits<double>::max(), std::numeric_limits<double>::max()}` (not infinity).

einola replied:
I'm not sure how that would work, because this constructor calls the ``ModelArray(const Type type)`` constructor. But using a default value is a good idea, so I removed this constructor and added a bounds argument to the existing ``ModelArray(const Type type)`` constructor, with ``{-std::numeric_limits<double>::max(), std::numeric_limits<double>::max()}`` as default.

timspainNERSC commented on core/src/include/ModelArray.hpp:
```diff
@@ -620,6 +634,20 @@ class ModelArray {
     static bool hasDoF(const Type type);
 
+private:
+    double lowerPhysicalLimit = -std::numeric_limits<double>::infinity();
+    double upperPhysicalLimit = std::numeric_limits<double>::infinity();
```
Possibly `std::numeric_limits<double>::max()`, rather than `std::numeric_limits<double>::infinity()`.

einola replied:
Done

timspainNERSC commented on core/src/ModelArray.cpp:
```diff
@@ -276,6 +274,40 @@ ModelArray::MultiDim ModelArray::locationFromIndex(Type type, size_t index)
+    // Now we check the bounds and set the array index (i) and value if we're out of bounds
+    size_t i;
+    double value;
+    if (masked.minCoeff() < lowerPhysicalLimit) {
+        value = masked.col(0).minCoeff(&i);
+    } else if (masked.maxCoeff() > upperPhysicalLimit) {
+        value = masked.col(0).maxCoeff(&i);
```
I would suggest changing the limits to a `<=` test. For example, if one were to apply strict limits to ice concentration, then the allowable values would be 0≤c_ice≤1, with both 0 and 1 as allowed values. To set this limit with the current implementation would require setting limits of `{-std::numeric_limits<double>::min(), 1.0+std::numeric_limits<double>::epsilon()}`. This has the knock-on effect of requiring the default limits to be `{-std::numeric_limits<double>::max(), std::numeric_limits<double>::max()}`, rather than using `std::numeric_limits<double>::infinity()`.

einola replied:
No, but I'll concede that the code is not very intuitive.

The ``if`` returns true if the value is _outside_ the allowed range, in which case we ask Eigen to find the offending value and its location and we proceed to throw an error.

So, with the current code, an initialisation of cice (for example) would be

```
cice(ModelArray::Type::DG, {0, 1})
```

This gives an allowed range of ``0 ≤ cice ≤ 1``, because as long as this is true, neither the ``if`` on line 299 nor the ``else if`` on line 301 is triggered, which brings us to a ``return`` on line 304.

timspainNERSC replied:
Probably should expand the comment at line 296 to explain this.

timspainNERSC commented on core/src/ModelArray.cpp:
```diff
@@ -276,6 +274,40 @@ ModelArray::MultiDim ModelArray::locationFromIndex(Type type, size_t index)
+    locStr.pop_back();
+    locStr.push_back(']');
+
+    throw std::runtime_error("Field contains out-of-bounds value(s), " + std::to_string(value)
+        + " not in [" + std::to_string(lowerPhysicalLimit) + ","
+        + std::to_string(upperPhysicalLimit) + "]. Error at index " + locStr + ".");
+}
```
Further to above comment, this error string implies that the acceptable range is inclusive of the limit values.

einola replied:
See above

timspainNERSC replied:
Review the comments to make sure this is explained.
(Also 'returned' is a perfectly reasonable non-jargon word that doesn't need an apostrophe.)

timspainNERSC commented on core/src/PrognosticData.cpp:
```diff
@@ -15,16 +15,23 @@
 
 namespace Nextsim {
 
+static const std::string checkFieldsKey = "debug.check_fields";
+static const std::string checkFieldsFastKey = "debug.check_fields_fast";
+static const std::map<int, std::string> keyMap
+    = { { PrognosticData::CHECKFIELDS_KEY, checkFieldsKey },
+          { PrognosticData::CHECKFIELDSFAST_KEY, checkFieldsFastKey } };
+static constexpr bool checkFieldsDefault = false;
+static constexpr bool checkFieldsFastDefault = true;
```
This should be moved to the `CheckingModelComponent` class. There would need to be a way for subclasses to provide their own configuration prefix and a call to `CheckingModelComponent::configure()` in the `configure()` function of each subclass.
`CheckingModelComponent`:
```
std::string configurationPrefix() virtual = 0;
…
configure()
{
…
    boolCheckAll() = Configured::getConfiguration( configurationPrefix() + checkAllKey, checkFieldsDefault);
    boolCheckFields
        = Configured::getConfiguration(configurationPrefix() + checkFieldsKey, checkFieldsFastDefault);
```
Also, common text for the `help-config` test could be provided by `CheckingModelComponent`, to be used by all subcalsses.

einola replied:
Yes, but I decided that the individual modules don't need to have a fine-grained control over their own checks. So, the user has two options now:

debug.check_fields: If true, all the fields registered in the ModelArray system are checked, as well as all the fields individual checking modules register. Default is false.
deebug.check_fields_fast: If true, a select few fields are checked by PrognosticData. Default true.

So, now only PrognosticData needs to have config keys and help.

timspainNERSC commented on core/src/PrognosticData.cpp:
```diff
@@ -51,6 +58,20 @@ void PrognosticData::configure()
+            fieldsToCheck.emplace_back(field.first, field.second);
+        }
+    } else if (boolCheckFields) {
+        fieldsToCheck.emplace_back("thickness", &hiceAdvection);
+        fieldsToCheck.emplace_back("concentration", &ciceAdvection);
+        fieldsToCheck.emplace_back("u_velocity", getStore().getArrayRef(Protected::ICE_U));
+        fieldsToCheck.emplace_back("v_velocity", getStore().getArrayRef(Protected::ICE_V));
```
Consider adding a function to the superclass that takes a `std::map<std::string, ModelArrayReference>` argument, rather than repeating `emplace_back` so many times in the derived class code.

einola replied:
Nice suggestion!

timspainNERSC reviewed (changes requested):
A lot of the common code could be moved into the `CheckingModelComponent` class, allowing the interface for checking the fields to be simpler in each of the derived classes and reducing the amount of boilerplate that needs to be introduced when a class is converted to become a `CheckingModelComponent`.

einola committed:
Remove initialisation by NaN

I put it there for debugging purposes and forgot to remove it again.

einola committed:
Set physical limits to +/- max, not inf

That makes more sense, and it's probably a better value to compare
against than IEEE inf. See comment from Tim on github.

einola committed:
Cleaning changes for ModelArray class

Moved the implementation of setLimits to the cpp file. Removed the
constructor with bounds and added the bounds argument, with default
values, to the existing ModelArray(const Type type) constructor.

einola committed:
Replace fieldsToCheck.emplace_back calls with a function call

Creates a function called addChecks which takes either a pair of string
and ModelArray pointer or a map of strings and ModelArray pointers.
These then get added to the fieldsToCheck vector, which is traversed by
checkFields, as before. See Tim's comment on github.

einola committed:
Do away with the overloaded addChecks function

The overload accepting a pair wasn't necessary, and I only used it once
in the code, so I got rid of it.

einola committed:
Remove fine-grained control on checking

Removes the option to check or not to check in individual modules. This
is not needed in general. Now, only PrognosticData decides if we do a
fast check (debug.check_fields_fast, true by default), or if everyone
checks everything (debug.check_fields, false by default). I also renamed
the boolean checking variables to reflect this change.

einola committed:
Correct python environment for mac tests

It should by python, not python3

einola committed:
Reset changes to Logged.cpp

These were just there for debugging purposes.

einola committed:
Reset changes to Iterator.hpp

I don't even know why they were there!

einola committed:
Reset changes to Model.hpp

It was only a date-change in the header!

einola committed:
Remove unused headers

They were lingering in ModelComponent.cpp

einola committed:
Revert useless changes

I'd been changing things back and forth, ending up with something
different from develop, but functionally equivalent. Reverting back to
develop now.

einola committed:
Reverting useless changes

Some formatting changes, compared to develop, snuck in. Now reverted.

einola committed:
Undo useless changes

Just leftovers from messing about with the TOPAZOcean class.

einola committed:
Revert "Reset changes to Logged.cpp"

This reverts commit 94369e4f934b08c168677f7b5b832835793f1114. It was
needed for propper debugging output.

einola committed:
Fix physical limits to 'evap'

It can also be negative, i.e. condensation.

einola commented:
Should be ok now, after removing functionality that I don't think is necessary! Have a look on Monday @timspainNERSC

einola committed:
Merge remote-tracking branch 'origin/develop' into issue774_check_fields

# Conflicts:
#	.github/workflows/test_suite.yml
#	core/src/Iterator.cpp
#	core/src/Logged.cpp
#	core/src/Model.cpp
#	core/src/ModelArray.cpp
#	core/src/ModelComponent.cpp
#	core/src/PrognosticData.cpp
#	core/src/include/ConfigurationHelp.hpp
#	core/src/include/ModelArray.hpp
#	core/src/include/ModelArrayReferenceStore.hpp
#	core/src/include/ModelComponent.hpp
#	core/src/include/PrognosticData.hpp
#	core/src/main.cpp
#	core/src/modules/DynamicsModule/include/DummyDynamics.hpp
#	dynamics/src/include/BrittleCGDynamicsKernel.hpp
#	physics/src/modules/AtmosphereBoundaryModule/ERA5Atmosphere.cpp
#	physics/src/modules/include/IAtmosphereBoundary.hpp
#	physics/test/IceGrowth_test.cpp
#	physics/test/ThermoWintonTemperature_test.cpp

timspainNERSC commented on core/src/include/ModelArray.hpp:
```diff
@@ -111,12 +111,15 @@ class ModelArray {
     ModelArray();
     /*!
-     * @brief Construct a ModelArray of the given type and name
+     * @brief Construct a ModelArray of the given type
```
Thanks for spotting and fixing this :)

timspainNERSC commented on core/src/PrognosticData.cpp:
```diff
@@ -50,6 +57,22 @@ void PrognosticData::configure()
+        addChecks({
+            { "thickness", &hice },
+            { "concentration", &cice },
+            { "u_velocity", getStore().getArrayRef(Protected::ICE_U) },
+            { "v_velocity", getStore().getArrayRef(Protected::ICE_V) },
+        });
```
The ice velocity components are not owned by `PrognosticData` and should not be set to be checked here. The call to `addChecks` should be called in `IDynamics` instead.

einola replied:
Yes, I see that, but this call is dependent upon the option ``debug.check_fields_fast`` being ``true``. So, I decided it would be better to do all of this in one place, where we have access to ``debug.check_fields_fast`` in the first place. Since ``IDynamics`` is not a ``Configured`` class, then I don't think it can have access to ``debug.check_fields_fast``, and I don't think making ``IDynamics`` ``Configured`` is a good idea.

So, my preferred solution is only to check thickness and concentration if ``debug.check_fields_fast`` is true. It's sufficient, because the idea is only to stop the model gracefully if everything goes wrong and errors in the velocity field will propagate to the thickness or concentration fields very quickly.

timspainNERSC commented on core/src/include/ModelArrayReferenceStore.hpp:
```diff
@@ -69,6 +69,20 @@ class ModelArrayReferenceStore {
         return dataMap;
     }
 
+    // For a given field name return the ModelArray pointer pointing to the data
+    const ModelArray* getArrayRef(const std::string& field)
+    {
+        try {
+            return storeRW.at(field);
+        } catch (std::out_of_range&) {
+            try {
+                return storeRO.at(field);
+            } catch (std::out_of_range&) {
+                throw;
+            }
+        }
+    }
+
```
Due to the way `ModelArrayReferenceStore` allows deferred registration, this is SIGSEGV bait. If ice u,v is check in their owning module (that is, the dynamics module), then this function is unnecessary.

einola replied:
Yes, that's a fair point. I'm removing this function.

timspainNERSC reviewed (changes requested):
Bypassing the designed method of getting data from `ModelArrayReferenceStore` is asking for trouble.

einola committed:
Merge branch 'develop' into issue774_check_fields

# Conflicts:
#	core/src/modules/include/SharedArrayNames.ipp
#	test/era5_topaz4_test_data.py

einola committed:
Better comments in checkLimits

The code is not very intuitive, and the comments weren't explaining it
very well. See also discussion on github.

einola committed:
Remove ModelArrayReferenceStore::getArrayRef and code using it

It's not a safe function because the the storeRW.at(field) or
storeRO.at (field) may not have been initialised (see Tim's comment on
github). I also removed the check for velocities in PrognosticData when
debug.check_fields_fast is true, because this depended on getArrayRef.
Removing this check is ok, because errors in the velocity field will
quickly propagate to the concentration and thickness fields, which are
still checked.

timspainNERSC reviewed (approved):
I've only reviewed the changes since I last viewed the PR, but I think I am happy for this to be merged now.

einola committed:
Merge branch 'develop' into issue774_check_fields

einola committed:
Remove bounds on fwFlux and sFlux

They are not well defined in a stand-alone setup; i.e. they may appear
unphysical large, but it's ok because the stand-alone is a bit strange
sometimes.
